galaxy_ng/app/api/ui/viewsets/collection.py

```diff
@@ -139,19 +142,30 @@ def get_queryset(self):
                 sign_state=Value("unsigned"),
             )
 
-        query_params = Q()
-        for collection_id, version in collection_versions.items():
-            query_params |= Q(collection_id=collection_id, version=version)
-
-        version_qs = CollectionVersion.objects.select_related("collection").filter(query_params)
-
-        base_total_qs = base_versions_query.filter(
-            namespace=OuterRef("namespace"), name=OuterRef("name")
+        # The main queryset to be annotated
+        version_qs = base_versions_query.select_related("collection")
+
+        # AAH-1484 - replacing `Q(collection__id, version)` with this annotation
+        # This builds `61505561-f806-4ddd-8f53-c403f0ec04ed:3.2.9` for each row.
+        # this is done to be able to filter at the end of this method and
+        # return collections only once and only for its highest version.
+        version_identifier_expression = Func(
+            F("collection__pk"), Value(":"), F("version"),
+            function="concat",
+            output_field=CharField(),
         )
 
         version_qs = version_qs.annotate(
             deprecated=Exists(deprecated_query),
-            **self.build_signing_annotations(base_total_qs)
+            version_identifier=version_identifier_expression,
+            **self.build_signing_annotations(base_versions_query)
```

Comment:
Why does this change from `base_total_qs` to `base_versions_query`?

Reply:
Reorganizing @ironfroggy I moved `base_total_qs` to be calculated inside the `build_signing_annotations` method, that `base_total_qs` queryset is used only to feed the `total: n` field which tells how many versions we have for each collection.